### DISCUSSION TITLE
SoQt -> Quarter

### DISCUSSION THREAD

**Discussion by sanguinariojoe** [SAP-archive/fedem-gui#21] [Q&A] [answered]
As far as I can see coin3d guys are promoting Quarter as a replacement for SoQt:

https://github.com/coin3d/soqt/blob/master/README#L12

Is a change FreeCAD already carried out. On top of that, I tested the mentor examples, and they all fail with soqt + qt6, while they work flawless with soxt. I also implemented some of them on quarter, and they work just fine.

So my plan now is makin possible to select between soqt and quarter

**Comment by sanguinariojoe** [comment]
I put this on Q&A, but it is more like an idea

**Comment by kmoks** [comment, marked as answer]
Yea, I've seen that the Quarter package is supposed to replace SoQt. But I'm not sure what effort it will take to do this change in the FEDEM GUI, so I've not considered doing this yet. But you are very welcome to try it out and report any findings you will like to share here.

I have now tried to build the GUI on Linux myself (on Ubuntu 22.04 with gcc 11.4). See the updated [BUILD.md](https://github.com/kmoks/fedem-gui/blob/qt6-port/BUILD.md#build-fedem-gui-application-on-linux) file on the qt6-port branch for details of this configuration.
I experience the same thing as reported in #17 when built for Release; it crashes immediately at startup. However, with `CMAKE_BUILD_TYPE=Debug`  it does not crash any more and the GUI lanches nicely, including the graphics window of Coin.

The crash for Release build is no good though, and need further investigation.

**Reply by kmokstad** [reply]
The Release build crash has now been resolved, see #25.

**Reply by sanguinariojoe** [reply]
Hey! Sorry, I had a though week!

I am checking it today or tomorrow!

**Reply by sanguinariojoe** [reply]
Thanks by the way!

**Comment by sanguinariojoe** [comment]
As I mentioned on #25 , effectivaly the qt6 port compiles and runs. Unfortunately I still need Quarter because the SoQt redering is broken.

**Reply by sanguinariojoe** [reply]
I think this week I can have a candidate

**Reply by kmoks** [reply]
Great. Thanks! Looking forward to see that :)

**Reply by sanguinariojoe** [reply]
Well, I already have the port. It is crashing obviously, but I feel I am close!

**Reply by sanguinariojoe** [reply]
All right!

![image](https://github.com/SAP/fedem-gui/assets/1668392/d6a1c8d6-ef38-49df-b187-ebf60e587932)

**Reply by sanguinariojoe** [reply]
I am making a PR so we can further discuss

**Reply by kmoks** [reply]
Yes, please make a PR against the kmoks/qt6-port branch and we can look at it closer. But it looks great! You should maybe test it with a real model. For instance, try to open sample model available at https://github.com/openfedem/public_models/tree/main/loader.

**Reply by sanguinariojoe** [reply]
I suspect github is not notifying you pull requests on your fork. Look, [this one](https://github.com/kmoks/fedem-gui/pull/1) flew totally under your radar, and the [Quarter one](https://github.com/kmoks/fedem-gui/pull/2) I opened it yesterday

**Reply by kmoks** [reply]
You're right, I was not notified by your low-hanging fruit PR either. Only saw it now when you pointed out. But I found your branch independently so I merged them into my branch. I will take a look at your new PR later today or tomorrow, and let you know. Details will be put as comments in kmoks/fedem-gui#2. Thanks a lot for doing this :)

**Reply by sanguinariojoe** [reply]
Just to feed a bit up, it seems the problem is that the FdMultiplyTransforms SoEngine is never evaluated, so all the stuff on FdQtViewer::processEvent() is actually working... It is just useless...

Now I just need to discover why...

**Reply by sanguinariojoe** [reply]
Well, the first problem was that `FdQtViewer::setCamera()` was never called. I fixed that, so I am sure the FdMultiplyTransforms is installed on the camera. It is still not working though

**Reply by sanguinariojoe** [reply]
I have created the following sandbox, and everything seems to work just fine there. What would be different on fedem-gui?

```
#include <QApplication>
#include <QWheelEvent>
#include <Quarter/Quarter.h>
#include <Inventor/SbColor.h>
#include <Inventor/nodes/SoBaseColor.h>
#include <Inventor/nodes/SoCone.h>
#include <Inventor/nodes/SoSeparator.h>
#include "FdQuarterViewer.H"
#include "FdMultiplyTransforms.H"
#include <iostream>

using namespace SIM::Coin3D::Quarter;

class FdQtViewer : public SoQtViewer
{
public:
  FdQtViewer(QWidget* parent  = NULL,
             const char* name = NULL,
             SbBool buildInsideParent = TRUE,
             SoQtViewer::Type type = BROWSER)
    : SoQtViewer(parent, name, buildInsideParent, type)
  {
    myMultiplier = new FdMultiplyTransforms;
    myMultiplier->ref();
  }
  virtual ~FdQtViewer()
  {
    myMultiplier->unref();
  }

  virtual void setCamera(SoCamera *cam)
  {
    if ( cam )
    {
      cam->orientation.connectFrom(&myMultiplier->rotation);
      cam->position.connectFrom(&myMultiplier->translation);
    }
    SoQtViewer::setCamera(cam);
  }

  virtual void processEvent( QEvent *qevent ) override
  {
    std::cout << "FdQtViewer::processEvent" << std::endl;
    bool eventHandeled = false;

    switch ( qevent->type() )
    {
      case QEvent::Wheel:
        std::cout << "\tQEvent::Wheel..." << std::endl;
        {
          QWheelEvent * wheelEv = (QWheelEvent *) qevent;
          std::cout << wheelEv->angleDelta().y() << std::endl;
          SbVec3f newCameraPosition = getCamera()->position.getValue() + SbVec3f(0, 0, wheelEv->angleDelta().y() / 120.0);
          this->myMultiplier->trans2 = newCameraPosition;
          auto v = getCamera()->position.getValue();
          std::cout << "\t" << v[0] << "," << v[1] << "," << v[2] << std::endl;
          eventHandeled = true;
        }
        break;
      default:
        eventHandeled = false;
        break;
    }

    // Pass upwards
    if ( !eventHandeled )
      this->SoQtViewer::processEvent(qevent);
  }

private:
  FdMultiplyTransforms * myMultiplier;
};

int
main(int argc, char ** argv)
{
  QApplication app(argc, argv);
  Quarter::init();

  FdMultiplyTransforms::init();

  SoSeparator * root = new SoSeparator;
  root->ref();
  SoBaseColor * col = new SoBaseColor;
  col->rgb = SbColor(1, 1, 0);
  root->addChild(col);
  root->addChild(new SoCone);

  FdQtViewer * viewer = new FdQtViewer();
  viewer->setSceneGraph(root);
  viewer->toggleCameraType();
  viewer->setBackgroundColor(SbColor(0.1, 0.1, 0.0));

  // Pop up the QuarterWidget
  viewer->show();
  // Loop until exit.
  app.exec();
  // Clean up resources.
  root->unref();
  delete viewer;
  Quarter::clean();
  return 0;
} 

```
